starknet_py/net/models/abi/definition.py

```diff
@@ -0,0 +1,180 @@
+        l1_handlers = grouped[L1_HANDLER_ENTRY]
+
+        if len(l1_handlers) > 1:
+            raise AbiDecodingError("L1 handler in ABI must be defined at most once")
+
+        if len(constructors) > 1:
+            raise AbiDecodingError("Constructor in ABI must be defined at most once")
+
+        structures = AbiDefinition._parse_abi_structs(structs_dict)
+        parser = TypeParser(structures)
+
+        return AbiDefinition(
+            defined_structures=structures,
+            constructor=AbiDefinition._parse_abi_function(constructors[0], parser)
+            if constructors
+            else None,
+            l1_handler=AbiDefinition._parse_abi_function(l1_handlers[0], parser)
+            if l1_handlers
+            else None,
+            functions={
+                name: AbiDefinition._parse_abi_function(entry, parser)
+                for name, entry in functions_dict.items()
+            },
+            events={
+                name: AbiDefinition._parse_event(entry, parser)
+                for name, entry in events_dict.items()
+            },
+        )
+
+    @staticmethod
+    def _parse_abi_structs(
+        structs_dict: Dict[str, StructDict]
+    ) -> Dict[str, StructType]:
+        # Contains sorted members of the struct
+        struct_members: Dict[str, List[StructMemberDict]] = {}
+        structs: Dict[str, StructType] = {}
+
+        # Example problem (with a simplified json structure):
+        # [{name: User, fields: {id: Uint256}}, {name: "Uint256", ...}]
+        # User refers to Uint256 even though it is not known yet (will be parsed next).
+        # This is why it is important to create the structure types first. This way other types can already refer to
+        # them when parsing types, even thought their fields are not filled yet.
+        # At the end we will mutate those structures to contain the right fields. An alternative would be to use
+        # topological sorting with an additional "unresolved type", so this flow is much easier.
+        for name, struct in structs_dict.items():
+            structs[name] = StructType(name, OrderedDict())
+            struct_members[name] = sorted(struct["members"], key=lambda m: m["offset"])
+
+        # Now parse the types of members and save them.
+        type_parser = TypeParser(structs)
+        for name, struct in structs.items():
+            members = AbiDefinition._parse_members(
+                f"members of structure '{name}'",
+                # pyright can't handle list of StructMemberDict here, even though TypedMemberDict
+                # is parent of StructMemberDict
+                cast(List[TypedMemberDict], struct_members[name]),
+                type_parser,
+            )
+            struct.types.update(members)
+
+        # All types have their members assigned now
+        return structs
+
+    @staticmethod
+    def _parse_abi_function(
+        function: FunctionDict, type_parser: TypeParser
```

Comment:
`TypeParser` is passed around a lot by all this methods, couldn't the structure be changed somehow so it doesn't have to be passed by methods? 

Perhaps having a separate AbiParser class with non static methods and having it as a member. I'm not a fan of chains of staticmethods if it can be avoided.

Reply:
Ok I thought about it for a while and I have an idea

Reply:
I prefer functional approach to this, but added AbiParser. Unfortunately parsing functions and events is dependant on parsing structures. I don't like late init variables in cases like this, but this was the simplest solution.

Reply:
Late init variable is not ideal, but I like the new approach a bit more than the previous one. 👍

Reply:
Happy to hear you like it <3